lib/chat/db/scope/key_scope.ex

```diff
@@ -132,47 +132,60 @@ defmodule Chat.Db.Scope.KeyScope do
       )
 
     acc_set
-    |> union_set(chunk_keys)
-    |> union_set(file_index)
-    |> union_set(file_chunks)
-    |> union_set(files)
-    |> union_set(memo_index)
-    |> union_set(memos)
-    |> union_set(room_invite_index)
-    |> union_set(room_invites)
+    |> Utils.union_set(chunk_keys)
+    |> Utils.union_set(file_index)
+    |> Utils.union_set(file_chunks)
+    |> Utils.union_set(files)
+    |> Utils.union_set(memo_index)
+    |> Utils.union_set(memos)
+    |> Utils.union_set(room_invite_index)
+    |> Utils.union_set(room_invites)
   end
 
   defp add_invitation_dialogs(acc_set, snap, pub_keys) do
-    dialog_keys =
-      snap
-      |> db_stream({:dialogs, 0}, {:"dialogs\0", 0})
-      |> Stream.filter(fn {_full_dilaog_key, %Dialog{a_key: a_key, b_key: b_key}} ->
-        MapSet.member?(pub_keys, a_key) or MapSet.member?(pub_keys, b_key)
-      end)
-      |> Stream.map(&just_keys/1)
-      |> Enum.to_list()
-      |> MapSet.new()
-
-    dialog_binkeys =
-      dialog_keys
-      |> Enum.map(fn {:dialogs, dialog_key} -> dialog_key end)
-      |> MapSet.new()
-
-    snap
-    |> db_stream({:dialog_message, 0, 0, 0}, {:"dialog_message\0", 0, 0, 0})
-    |> Stream.filter(fn {{:dialog_message, key, _, _}, %Message{type: type}} ->
-      type == :room_invite and MapSet.member?(dialog_binkeys, key)
-    end)
-    |> Stream.map(&just_keys/1)
-    |> union_set(dialog_keys)
-    |> union_set(acc_set)
+    dialogs = Utils.get_dialogs(snap)
+
+    {checkpoints_dialog_keys, checkpoints_invitation_messages} =
+      extract_dialogs_with_invitations(:checkpoints, snap, dialogs, [pub_keys, nil])
+
+    operator_key =
+      Utils.get_invitation_sender_key(:sender, dialogs, checkpoints_invitation_messages)
+
+    {operator_dialog_keys, operator_invitation_messages} =
+      extract_dialogs_with_invitations(:user, snap, dialogs, [
+        operator_key,
+        checkpoints_dialog_keys
+      ])
+
+    nested_user_key =
+      Utils.get_invitation_sender_key(:recipient, dialogs, operator_invitation_messages)
+      |> MapSet.difference(pub_keys)
+
+    {nested_dialog_keys, nested_invitation_messages} =
+      extract_dialogs_with_invitations(:user, snap, dialogs, [
+        nested_user_key,
+        operator_dialog_keys
+      ])
+
+    {
+      [checkpoints_invitation_messages, operator_invitation_messages, nested_invitation_messages]
+      |> Stream.concat()
+      |> Stream.map(&Utils.just_keys/1)
+      |> union_set_dialog_keys([
+        checkpoints_dialog_keys,
+        operator_dialog_keys,
+        nested_dialog_keys
+      ])
+      |> Utils.union_set(acc_set),
+      MapSet.union(pub_keys, nested_user_key)
```

Comment:
It is hard to understand what happens here

may be start with comments

Reply:
each step commented